### PR TITLE
Reorg + JIT Document building

### LINKED ISSUE #824: U - Yari Architecture documentation
## Acceptance Criteria

- [ ] A high-resolution document, detailing the different architectures, toolchains, running processes, working modes (local versus building for production), various watching/caching layers and the user/writer needs they serve completed

### LINKED ISSUE #832: ListSubPages macro doesn't pick up changes
`ListSubPages` macro doesn't pick up changes (e.g., when @Elchi3 added the new “Firefox 79” page).

Steps to reproduce:

1. Go to http://localhost:3000/en-US/docs/Mozilla/Firefox/Releases/78 (or whatever is latest)
2. Use the Toolbar to create the `79` page. 
3. Visit http://localhost:3000/en-US/docs/Mozilla/Firefox/Releases/

### LINKED ISSUE #976: /en-US/search-index.json isn't working
Note, this is happening in the reductio-reloaded branch

Steps to reproduce:

1. `yarn start`
2. Go to http://localhost:3000/en-US/search-index.json
3. Click into (hover over) the search input and this happens:
Screen Shot 2020-07-28 at 3 58 12 PM

There are two problems here. First of all, it shouldn't say "Initializing index..." when it actually failed. Also, the placeholder still says "Go ahead. Type your search..."

The reason it's failing in the first place is that, in `yarn start` is because of this:
```
▶ curl -v http://localhost:5000/en-US/search-index.json
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 5000 (#0)
> GET /en-US/search-index.json HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Type: application/json; charset=utf-8
< Date: Tue, 28 Jul 2020 20:01:47 GMT
< Connection: keep-alive
< Content-Length: 0
<
* Connection #0 to host localhost left intact
* Closing connection 0

```
which means it returned (immediately) with an empty body. 
We should instead simulate what would happen in production which is that it should simply wait until it's ready.

### PR DESCRIPTION
**Pull request by Gregoor**
The long awaited sequel to #762 

This time with a new feature: It fixes #832, fixes #824 

Fixes #976

Alrighty, you two. The framework is in place. Here's my idea for the way forward (ofc feel free to make other suggestions):

I'd work on (re)adding
- [x] sitemaps
- [x] live samples
  - [x] live sample building
- [x] make the functional tests pass
- [x] CRUD
- [x] Document watcher
- [x] updating documentation
- [x] support additional content root (aka archive)
- [x] build search index json
- [x] memoize and make it fast
- [x] `yarn build` should exit non-zero if an error happens
- [x] [search should wait to response on XHR requests](https://github.com/mdn/yari/issues/976)
___

@peterbe & @escattone: These are some other bigger remaining TODOs, I don't have a clear grasp on who'd be best for which one. But if you want to tell me which one(s) you want to take on, I'd create some sprint task issues to satisfy the sprint gods.

## ~~translations~~
~~As we three talked about the current lookup model doesn't work for reductio (`translation_of` living in the translated document, hence an O(1) lookup not being possible). It sounded like we were all vaguely in favor of having something like a `translations.json` in the locale root (just like `wiki_history.json`). But I don't want to speak for anyone else if there's still disagreement here.
Anyway, if there isn't disagreement. This would require a change to the importer (which now lives in its own workspace)~~

**Update:** Ryan and I talked about this yesterday, he made the good point that we don't know how to move forward with translations in general, so we shouldn't block on this. We know how to solve this in a reductio world but the effort might not be worth it at the moment.

## flaws
I think I broke some things there and I haven't backported (err frontported?) fixable flaws yet. Also happy to take this on as I broke it, but if you feel inclined to, that's also great.

## everything KS
I touched a lot of that code and haven't exhaustively checked the results yet. After a general review is done, I'd wanna compare output of the builder on `master` with this branch's output.

## things that were lost
Not on my todo list are stumptown content support and configurable building. My thinking is that the former requires a different model (i.e. indexing) and we haven't made sure whether we broke it at some point in the last while anyway. The latter seemed more useful in the "we're indexing everything" world. For me the added complexity doesn't have benefits that are obvious to me in the reductio world.
Anyway, I took quite the chainsaw swing at the builder, so I think it's quite likely that I missed some nuance here and there and dropped too much. So if you could help me / negotiate with me figure out what we need to bring back, that'd be helpful.

___

That should be the outline of what's left for feature parity. There's also performance parity, on my machine the full build is around 50% slower atm. I don't worry about this too much atm, as I don't think there's any optimization we did before that we can't do now. Just want to wait til I'm sure about the output being correct before I dive more into it.

### CODE REVIEW

**Comment by peterbe** [comment]
I look forward to learning from this. 
But, quick question; why do you leave out the "whole build-mode complexity"? Without a performant builder does it have any hope at all if we're going to jamstack deploy this?

**Comment by Gregoor** [comment]
I've put that in misleading way! By not bother with building I meant just using the builder as is (which requires some reverting of things I threw away and sharing of code between the two builders).

File: content/utils.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+function memoizeDuringBuild(fn) {
+  if (process.env.NODE_ENV === "development") {
+    return fn;
+  }
+
+  const cache = new Map();
+  return (...args) => {
+    const key = JSON.stringify(args);
```
Comment (escattone) [inline content/utils.js:37]: If multiple calls to `fn` include objects in `args` but don't use the same key order for those objects, then the memoization will be replicated. It might not be worth making this more robust in its handling of objects, but just something to keep in mind

File: content/document.js [hunk 2]
```diff
@@ -206,16 +193,42 @@ function update(contentRoot, folder, rawHtml, metadata) {
...
-function del(contentRoot, folder) {
-  const { metadata } = read(contentRoot, folder);
+function del(folder) {
+  const { metadata } = read(folder, { metadata: true });
   fs.rmdirSync(folder, { recursive: true });
-  updateWikiHistory(path.join(contentRoot, metadata.locale), metadata.slug);
+  updateWikiHistory(path.join(CONTENT_ROOT, metadata.locale), metadata.slug);
+}
+
+const findByURL = memoizeDuringBuild((url, fields = null) => {
```
Comment (escattone) [inline content/document.js:202]: This function has to be more robust in terms of its handling of the incoming `url`, since the incoming `url` can come from the arguments of macro calls (which are notoriously unreliable). It'll have to do two things:
- repair an unorthodox URL (like we do currently with https://github.com/mdn/yari/blob/674e331d2f79f3c340339f31cf80b387b2bf7e60/content/scripts/build.js#L244)
- replace a redirected URL with its new target URL

so basically it'll have to do what the current [`cleanUri`](https://github.com/mdn/yari/blob/674e331d2f79f3c340339f31cf80b387b2bf7e60/content/scripts/build.js#L369) does.

File: kumascript/index.js [hunk 3]
```diff
@@ -1,79 +1,54 @@
...
-        live_samples: { base_url: this.liveSamplesBaseUrl },
+async function renderFromURL(url) {
+  const { rawHtml, metadata } = Document.findByURL(url).document;
+  const [renderedHtml, errors] = await renderMacros(
+    rawHtml,
+    new Templates(),
+    {
+      ...{
+        path: url,
+        url: `${"this.options.sitemapBaseUrl"}${url}`,
```
Comment (escattone) [inline kumascript/index.js:25]: I suppose we should create a new constant (maybe just `SITE_URL`?) that can be overridden from the environment for this? Basically it would be the domain we're deploying to (and its protocol), so for example `https://developer.mozilla.org` or `https://pr9999.yari.io` or `http://localhost:5000` or something like that.

File: kumascript/src/api/wiki.js [hunk 4]
```diff
@@ -69,7 +69,7 @@ module.exports = {
       return html;
     }
 
-    let result = this.info.getResultFromCache(path);
+    let result = await this.renderFromURL(path);
     const pathDescription = this.info.getDescription(path);
 
     if (!result) {
```
Comment (escattone) [inline kumascript/src/api/wiki.js:75]: With your new approach, this entire `if` block is no longer needed.

File: content/document.js (hunk 2 above)
Comment (escattone) [reply content/document.js:202]: Oh, I see you already have this functionality within `info.js` as `return Redirect.resolve(repairURL(uri));` within `getUriKey()`. I wonder if `repairURL(uri)` belongs better in `content/utils.js`?

Comment (escattone) [reply content/document.js:202]: Actually, the more I think about this, we only really need to clean the URL's arriving from the content (e.g., arguments to macro calls), so I suppose we could just pre-wrap the `globals.renderFromURL` within `kumascript/src/environment.js`, so something like:
```js
globals.renderFromURL = (url) => renderFromURL(getUriKey(url));
```

File: kumascript/src/info.js [hunk 5]
```diff
@@ -162,48 +68,81 @@ class AllPagesInfo {
...
+    return [];
+  },
 
   getPage(url) {
-    const uriKey = this.getUriKey(url);
-    if (!this.pagesByUri.has(uriKey)) {
-      // Many of the macros expect an empty object
-      // when the requested document doesn't exist.
-      return {};
+    const result = Document.findByURL(url, { metadata: true });
```
Comment (escattone) [inline kumascript/src/info.js:124]: You'll need to clean the incoming `url`, so:
```suggestion
    const result = Document.findByURL(getUriKey(url), { metadata: true });
```

File: kumascript/src/info.js [hunk 6]
```diff
@@ -1,149 +1,55 @@
...
+    // we're only after the path of any incoming "url", but it's required by
+    // the URL constructor when the incoming "url" is relative.
+    const uri = new URL(url, BASE_URL).pathname
+      .replace(/\/$/, "")
+      .toLowerCase();
+    return Redirect.resolve(repairURL(uri));
+  },
 
   getDescription(url) {
     const uriKey = this.getUriKey(url);
```
Comment (escattone) [inline kumascript/src/info.js:55]: Is it safer to use `info.getUriKey` instead?

File: kumascript/src/info.js [hunk 7]
```diff
@@ -162,48 +68,81 @@ class AllPagesInfo {
...
+    return {
+      url: document.url,
+      locale,
+      slug,
+      title,
+      summary,
+      tags: tags || [],
+      translations: [], //TODO Object.freeze(buildTranslationObjects(data)),
+      get subpages() {
+        return Document.findChildren(url, { metadata: true }).map((document) =>
```
Comment (escattone) [inline kumascript/src/info.js:140]: Probably better to use `document.url` here since it has already been "cleaned":
```suggestion
        return Document.findChildren(document.url, { metadata: true }).map((document) =>
```

File: content/document.js [hunk 8]
```diff
@@ -206,16 +193,42 @@ function update(contentRoot, folder, rawHtml, metadata) {
...
+  return document ? { contentRoot: CONTENT_ROOT, folder, document } : null;
+});
+
+function findChildren(url, fields = null) {
+  const folder = urlToFolderPath(url);
+  const childPaths = glob.sync(
+    path.join(CONTENT_ROOT, folder, "*", HTML_FILENAME),
+    {
+      ignore: path.join(CONTENT_ROOT, getHTMLPath(folder)),
+    }
```
Comment (escattone) [inline content/document.js:216]: I don't think this is needed, at least it worked for me without it.

File: kumascript/src/environment.js [hunk 9]
```diff
@@ -133,6 +138,7 @@ class Environment {
     globals.page = globals.Page = freeze(page);
     globals.env = globals.Env = freeze(env);
     globals.info = freeze(info);
+    globals.renderFromURL = renderFromURL;
```
Comment (escattone) [inline kumascript/src/environment.js:141]: Since this function will receive URL's from the "wilds" of the content, it's important to clean the incoming URL's and maybe it's best to do that here?
```suggestion
    globals.renderFromURL = (url) => renderFromURL(info.getUriKey(url));
```

File: kumascript/index.js (hunk 3 above)
Comment (Gregoor) [reply kumascript/index.js:25]: Good idea, this was a bit of an easy to overlook placeholder for something that still needs to be worked on, thanks for highlighting it!

File: content/utils.js (hunk 1 above)
Comment (Gregoor) [reply content/utils.js:37]: Yup, good point! I also considered explicitly having a cache key getter as a parameter, but the call sites were simple enough that it seemed fine. I'll add a comment though, explaining the limitation.

File: kumascript/index.js (hunk 3 above)
Comment (Gregoor) [reply kumascript/index.js:25]: Out of curiousity: Why do we want to use an absolute URL here? Couldn't we just as well skip the host?

File: kumascript/src/api/wiki.js (hunk 4 above)
Comment (Gregoor) [reply kumascript/src/api/wiki.js]: Awesome, removed! Thanks

File: kumascript/src/info.js (hunk 5 above)
Comment (Gregoor) [reply kumascript/src/info.js:124]: Fixed, thank you!

File: content/document.js (hunk 8 above)
Comment (Gregoor) [reply content/document.js:233]: Indeed!

File: kumascript/src/info.js (hunk 6 above)
Comment (Gregoor) [reply kumascript/src/info.js]: Yes, much easier on the eyes. Thanks for the suggestion

**Comment by Gregoor** [comment]
Alrighty, you two. The framework is in place. Here's my idea for the way forward (ofc feel free to make other suggestions):

I'd work on (re)adding
- [ ] sitemaps
- [ ] make the functional tests pass
- [ ] CRUD
- [ ] Document watcher

___

@peterbe & @escattone: These are some other bigger remaining TODOs, I don't have a clear grasp on who'd be best for which one. But if you want to tell me which one(s) you want to take on, I'd create some sprint task issues to satisfy the sprint gods.

## translations
As we three talked about the current lookup model doesn't work for reductio (`translation_of` living in the translated document, hence an O(1) lookup not being possible). It sounded like we were all vaguely in favor of having something like a `translations.json` in the locale root (just like `wiki_history.json`). But I don't want to speak for anyone else if there's still disagreement here.
Anyway, if there isn't disagreement. This would require a change to the importer (which now lives in its own workspace)

## flaws
I think I broke some things there and I haven't backported (err frontported?) fixable flaws yet. Also happy to take this on as I broke it, but if you feel inclined to, that's also great.

## everything KS
I touched a lot of that code and haven't exhaustively checked the results yet. After a general review is done, I'd wanna compare output of the builder on `master` with this branch's output.

## things that were lost
Not on my todo list are stumptown content support and configurable building. My thinking is that the former requires a different model (i.e. indexing) and we haven't made sure whether we broke it at some point in the last while anyway. The latter seemed more useful in the "we're indexing everything" world. For me the added complexity doesn't have benefits that are obvious to me in the reductio world.
Anyway, I took quite the chainsaw swing at the builder, so I think it's quite likely that I missed some nuance here and there and dropped too much. So if you could help me / negotiate with me figure out what we need to bring back, that'd be helpful.

___

That should be the outline of what's left for feature parity. There's also performance parity, on my machine the full build is around 2x slower atm. I don't worry about this too much atm, as I don't think there's any optimization we did before that we can't do now. Just want to wait til I'm sure about the output being correct before I dive more into it.

**Comment by Gregoor** [comment]
Oh updating the docs is another thing I want to do, which will mainly be subtractive work. All the commands are still the same, there'd probably just be less of them.

**Comment by Gregoor** [comment]
Okay things are a little intertwined: to fix the remaining tests, I need to fix flaws first 🙃 

Also had the audacity to comment out some tests that checked for `modified` and `popularity` being present. I'm wondering whether we'd want to change those tests to check for surface area instead of internal data structures anyway? `modified` is not used anywhere yet though, with popularity we could check for the order of the suggested items (though to a degree we're also at the mercy of our search indexing library there). Wdyt @peterbe?

**Comment by peterbe** [comment]
> Also had the audacity to comment out some tests that checked for `modified` and `popularity` being present. I'm wondering whether we'd want to change those tests to check for surface area instead of internal data structures anyway? `modified` is not used anywhere yet though, with popularity we could check for the order of the suggested items (though to a degree we're also at the mercy of our search indexing library there). Wdyt @peterbe?

Quick drive-by comment; The `modified` is used in the `kumastyles` branch. At some point, we should check that presence with headless tests, like a human QA would do. 
Also, the popularity *number* is never important. So yes, we can totally look at the order they appear. 
Headless tests are always better than "jest tests" but some things can weird/tricky to test with headless tests. But what always matters the most is that the functionality is there.

**Comment by escattone** [comment]
@Gregoor @peterbe I'll start by taking `Everything KS` and the KS-related flaws if that's cool with you guys.

**Comment by peterbe** [comment]
Clicking "Edit in your editor" fails with:
Screen Shot 2020-07-21 at 12 27 40 PM

The in-browser editor fails with:
```
12:28:36 PM server.1   |  (node:81307) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'metadata' of null
12:28:36 PM server.1   |      at Object.update (/Users/peterbe/dev/MOZILLA/MDN/yari/content/document.js:150:28)
12:28:36 PM server.1   |      at /Users/peterbe/dev/MOZILLA/MDN/yari/server/document.js:34:14
12:28:36 PM server.1   |      at Layer.handle [as handle_request] (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/layer.js:95:5)
12:28:36 PM server.1   |      at next (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/route.js:137:13)
12:28:36 PM server.1   |      at withDocument (/Users/peterbe/dev/MOZILLA/MDN/yari/server/document.js:24:3)
12:28:36 PM server.1   |      at Layer.handle [as handle_request] (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/layer.js:95:5)
12:28:36 PM server.1   |      at next (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/route.js:137:13)
12:28:36 PM server.1   |      at Route.dispatch (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/route.js:112:3)
12:28:36 PM server.1   |      at Layer.handle [as handle_request] (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/layer.js:95:5)
12:28:36 PM server.1   |      at /Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/index.js:281:22
12:28:36 PM server.1   |  (node:81307) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 8)
```

**Comment by Gregoor** [comment]
Thanks Peter I fixed this in one go with CRUD.

Also I'm treating [my comment from above](https://github.com/mdn/yari/pull/881#issue-447610018) as the TODO list for this to land (and I moved it into the first post). There's an udpate to translations in there from Ryan and I talking about it yesterday.

**Comment by Gregoor** [comment]
Thanks for all the effort getting this to parity! We're close. Today I've put live samples and web-socket based change notifications in. For both I'd appreciate a check whether it generally makes sense. For the former probably from @escattone, for the latter from @peterbe.

Also I could use some guidance wrt merging master into this branch. Anything I should look out for when merging kumascript?

**Comment by escattone** [comment]
@Gregoor Awesome, I'll check out the live samples. Regarding guidance on how to merge master into this branch, for the `kumascript` workspace I would say that we always take `reductio-reloaded` rather than what's in`master`, but I'll double check that.

**Comment by Gregoor** [comment]
I'm a bit confused, the headless and search tests run fine for me. A complete test run yields success for 13/16. Whereas in GH action we only have 8/16. They all seem to be related to timeouts which I don't have an explanation for yet.

**Comment by escattone** [comment]
@Gregoor I think I found the problem. The `buildLiveSamplePageFromURL()` function within the preview server always gets lower-cased URL's, so when it extracts the live-sample ID from the URL, that's always lower-case as well. However, the sample D's are case sensitive, so the live-samples with ID's containing upper-case letters are never found. More tommorrow.

**Comment by escattone** [comment]
@Gregoor The fix is https://github.com/Gregoor/yari/pull/13.
